forc-util/src/fs_locking.rs

```diff
@@ -0,0 +1,197 @@
+use crate::{hash_path, user_forc_directory};
+use std::{
+    fs::{create_dir_all, remove_file, File},
+    io::{self, Read, Write},
+    path::{Path, PathBuf},
+};
+
+/// Very simple AdvisoryPathMutex class
+///
+/// The goal of this struct is to signal other processes that a path is being used by another
+/// process exclusively.
+///
+/// This struct will self-heal if the process that locked the file is no longer running.
+pub struct PidFileLocking(PathBuf);
+
+impl PidFileLocking {
+    pub fn new<X: AsRef<Path>, Y: AsRef<Path>>(
+        filename: X,
+        dir: Y,
+        extension: &str,
+    ) -> PidFileLocking {
+        let file_name = hash_path(filename);
+        Self(
+            user_forc_directory()
+                .join(dir)
+                .join(file_name)
+                .with_extension(extension),
+        )
+    }
+
+    /// Create a new PidFileLocking instance that is shared between the LSP and any other process
+    /// that may want to update the file and needs to wait for the LSP to finish (like forc-fmt)
+    pub fn lsp<X: AsRef<Path>>(filename: X) -> PidFileLocking {
+        Self::new(filename, ".lsp-locks", "dirty")
+    }
+
+    /// Checks if the given pid is active
+    #[cfg(not(target = "windows"))]
+    fn is_pid_active(pid: usize) -> bool {
+        use std::process::Command;
+        let output = Command::new("ps")
+            .arg("-p")
+            .arg(pid.to_string())
+            .output()
+            .expect("Failed to execute ps command");
```

Comment:
We can use the [sysinfo](https://docs.rs/sysinfo/latest/sysinfo/struct.System.html#method.processes) library for this

Reply:
Might be good to add a code comment to that effect, like
```
                // Not using sysinfo here because it has compatibility issues with fuel.nix
                // https://github.com/FuelLabs/fuel.nix/issues/64
```